File: LeetCode/0016. 3Sum Closest/3SumClosest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
 var threeSumClosest = function(nums, target) {
  nums.sort((a, b) => a - b)
  let minDiff = Number.MAX_SAFE_INTEGER, res
  for (let i = 0; i < nums.length; i++) {
      let left = i + 1, right = nums.length - 1
      while (left < right) {
          let tmp = nums[left] + nums[i] + nums[right]
          if (Math.abs(tmp - target) < minDiff) {
              res = tmp
              minDiff = Math.abs(tmp - target) 
          }
              
          if (tmp === target) {
              return target
          }
          else if (tmp < target)
              left++
          else
              right--
      }
  }
  return res
};

File: LeetCode/0016. 3Sum Closest/3SumClosest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        
        ans = inf

        for i in range(len(nums)): 
            if i and nums[i-1] == nums[i]: continue
            lo, hi = i+1, len(nums)-1
            while lo < hi: 
                sm = nums[i] + nums[lo] + nums[hi] 
                ans = min(ans, sm, key=lambda x: abs(x-target))
                if sm < target: lo += 1
                elif sm == target: return ans
                elif sm > target: hi -= 1

        return ans 